scheduler/core/programprovider/gpp/gppprogramprovider.py

```diff
@@ -0,0 +1,1333 @@
+from scheduler.core.sources.sources import Sources
+from scheduler.services import logger_factory
+
+
+__all__ = [
+    'GppProgramProvider',
+    'ocs_program_data',
+]
+
+logger = logger_factory.create_logger(__name__)
+
+
+DEFAULT_OCS_DATA_PATH = Path(ROOT_DIR) / 'scheduler' / 'data' / 'programs.zip'
+DEFAULT_PROGRAM_ID_PATH = Path(ROOT_DIR) / 'scheduler' / 'data' / 'program_ids.txt'
+
+
+def ocs_program_data(program_list: Optional[bytes] = None) -> Iterable[dict]:
+    try:
+        # Try to read the file and create a frozenset from its lines
+        if program_list:
+            list_file = program_list
+        else:
+            list_file = DEFAULT_PROGRAM_ID_PATH
+
+        if isinstance(program_list, bytes):
+            file = program_list.decode('utf-8')
+            id_frozenset = frozenset(f.strip() for f in file.split('\n') if f.strip() and f.strip()[0] != '#')
+        else:
+            with list_file.open('r') as file:
+                id_frozenset = frozenset(line.strip() for line in file if line.strip() and line.strip()[0] != '#')
+    except FileNotFoundError:
+        # If the file does not exist, set id_frozenset to None
+        id_frozenset = None
+    return read_ocs_zipfile(DEFAULT_OCS_DATA_PATH, id_frozenset)
+
+
+def read_ocs_zipfile(zip_file: str | PathLike[str], program_ids: Optional[FrozenSet[str]] = None) -> Iterable[dict]:
+    """
+    Since for OCS we will use a collection of extracted ODB data, this is a
+    convenience method to parse the data into a list of the JSON program data.
+    """
+    with zipfile.ZipFile(zip_file, 'r') as zf:
+        for filename in zf.namelist():
+            program_id = Path(filename).stem
+            if program_ids is None or program_id in program_ids:
+                with zf.open(filename) as f:
+                    contents = f.read().decode('utf-8')
+                    logger.info(f'Adding program {program_id}.')
+                    yield json.loads(contents)
+            else:
+                logger.info(f'Skipping program {program_id} as it is not in the list.')
+
+
+def parse_preimaging(sequence: List[dict]) -> bool:
+
+    preimaging = False
+    try:
+        if sequence[0][GppProgramProvider._AtomKeys.PREIMAGING].upper() == 'YES':
+            preimaging = True
+    except (KeyError, IndexError):
+        pass
+
+    return preimaging
+
+
+class GppProgramProvider(ProgramProvider):
+    """
+    A ProgramProvider that parses programs from the GPP Observing Database.
+    """
+
+    _GPI_FILTER_WAVELENGTHS = {'Y': 1.05, 'J': 1.25, 'H': 1.65, 'K1': 2.05, 'K2': 2.25}
+    _NIFS_FILTER_WAVELENGTHS = {'ZJ': 1.05, 'JH': 1.25, 'HK': 2.20}
+    _CAL_OBSERVE_TYPES = frozenset(['FLAT', 'ARC', 'DARK', 'BIAS'])
+
+    _site_for_inst = {'GMOS_NORTH': Site.GN, 'GMOS_SOUTH': Site.GS}
+
+    # GPP GMOS built-in GPU name to barcode
+    # ToDo: Eventually this needs to come from another source, e.g. Resource, ICTD, decide whether to use name or barcode
```

Comment:
I guess  Resource will take all the functions of ictd eventually

Reply:
yes